050-Kafka-Node-Pools.md

```diff
@@ -0,0 +1,821 @@
+
+### Stretch clusters
+
+There seems to be increased demand Kafka clusters stretched across multiple Kubernetes clusters.
+Apache Kafka itself has latency limitations and might not run well when stretched over distant clusters with big latency.
+But there are many use cases where stretched clusters might be useful.
+For example:
+
+* When running closely colocated Kubernetes clusters
+* When migrating from old Kubernetes cluster to new Kubernetes cluster
+
+## Proposal
+
+This proposal suggests introduction of Kafka _node pools_.
+_Node pools_ are groups of Kafka nodes which will have the same configuration.
+A Kafka cluster can consist of one or more pools.
+Each node pool can belong to only one Kafka cluster.
+
+### `KafkaNodePool` custom resource
+
+A new custom resource called `KafkaNodePool` will be added to Strimzi.
+One `KafkaNodePool` resource will represent one _node pool_.
+It will take over some of the configuration from the `Kafka` CR, reuse other parts from it and add some new configuration options as well.
+The `KafkaNodePool` resource will be a _public_ resource.
+It will be created and managed directly by users (unlike for example the `StrimziPodSet` resource which is used only internally by Strimzi and should not be touched by the users).
+
+The _cluster-wide_ configuration options will remain part of the `Kafka` CR.
+While the _pool-specific_ options will be taken from `KafkaNodePool` resources
+
+For the initial implementation, the following options are proposed for the `KafkaNodePool` resource and its `.spec` section:
+
+* `replicas` indicating the number of nodes in the pool
+* `storage` indicating the storage of the nodes in this pool
+* `resources` indicating the amount of resources (requests / limits) assigned to the nodes in this pool
+* `template` for customizing the pods and other resources which are part of the nodes in this pool
+* `jvmOptions` for customizing the JVM configuration
+
+For example:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: my-pool
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  resources:
+    requests:
+      memory: 8Gi
+      cpu: 2
+    limits:
+      memory: 8Gi
+      cpu: 4
+  jvmOptions:
+    -Xms: 2048m
+    -Xmx: 2048m
+  storage:
+    type: jbod
+    volumes:
+      - id: 0
+        type: persistent-claim
+        size: 100Gi
+        deleteClaim: true
+  template:
+    podSet:
+      metadata:
+        labels:
+          key1: label1
+    pod:
+      terminationGracePeriodSeconds: 60
+```
+
+The replicas and storage sections will be required in `KafkaNodePool`.
+The other fields will be optional.
+At the same time, the replicas and storage sections will be made optional in the `Kafka` resource since when using node pools, and they will ignored if set.
```

Comment:
I think there must be some words missing in "since when using node pools, and they will ignored"

Reply:
I have reworked this section.